### ISSUE TITLE
Event Listener and Report output incorrect or nunit2 test runs

### ISSUE THREAD

**Issue by owenniblock** [nunit/nunit-v2-framework-driver#15]
When I'm running nunit2 tests using the nunit v2 framework driver - the console outputs the correct number of passes and failures, but the XML sent to both the Event Listener and the xml result output is incorrect.

For example:
I have 3 tests which output:
Test Run Summary
  Overall result: Failed
  Test Count: 3, Passed: 1, Failed: 2, Warnings: 0, Inconclusive: 0, Skipped: 0
    Failed Tests - Failures: 1, Errors: 1, Invalid: 0

The TestResults.xml file reports:

```xml
<test-run id="2" testcasecount="3" result="Failed" total="3" passed="0" failed="0" inconclusive="0" skipped="0" asserts="0" engine-version="3.7.0.0" clr-version="4.0.30319.42000" start-time="2017-07-19 12:06:05Z" end-time="2017-07-19 12:09:15Z" duration="190.467029">
```

The Event Listener XML is the same:

```xml
<test-run id="2" testcasecount="3" result="Failed" total="3" passed="0" failed="0" inconclusive="0" skipped="0" asserts="0" engine-version="3.6.1.0" clr-version="4.0.30319.42000" start-time="2017-07-18 14:12:58Z" end-time="2017-07-18 14:16:08Z" duration="190.485957">
```

Interestingly, if I use the legacy nunit2 report format, the TestResults.xml file outputs correctly:

```xml
<test-results name="D:\MS\test-apps\Xamarin.NUnit2.Tests\bin\Debug\Xamarin.NUnit2.Tests.dll" total="3" errors="1" failures="1" not-run="0" inconclusive="0" ignored="0" skipped="0" invalid="0" date="2017-07-19" time="12:44:44">
```

Let me know if this issue is better placed in the main nunit3 repo - I'll have a dig about in the code and see if I can work out what's going on.

**Comment by CharliePoole** [comment]
I edited your description so your XML displays.

**Comment by CharliePoole** [comment]
Looking at the filename in your last example, I gather these are xamarin tests. If that's the case, the NUnit engine (and hence the console runner) would not know how to run them. Can you clarify how you are running the tests?

**Comment by owenniblock** [comment]
These are just standard tests (I'm working at Xamarin hence the namespace). For example:

```csharp
        [Test]
        public void SimpleSuccessTest()
        {
            Console.WriteLine("SimpleSuccessTest BEFORE WAIT:: " + DateTime.Now.ToString("yyyy-MM-dd hh:mm:ss"));

            Thread.Sleep(10000);

            Console.WriteLine("SimpleSuccessTest AFTER WAIT:: " + DateTime.Now.ToString("yyyy-MM-dd hh:mm:ss"));

            Assert.IsTrue(true);
        }
```

I'm running the tests using the console:
```
nunit3-console "D:\\MS\\test-apps\\Xamarin.NUnit2.Tests\\bin\\Debug\\Xamarin.NUnit2.Tests.dll"  --workers=0 --process=Single
```

**Comment by CharliePoole** [comment]
The results are interesting. The nunit2 format output is actually created from the nunit3 output. The nunit3 output is generated by the extension from the v2 object model. That suggests an error in creating the nunit3 output followed by a compensating error in reformatting to nunit2. Like a perfect storm of errors.

**Comment by owenniblock** [comment]
Interesting - so the correct data must be in there somewhere :) Sounds like it's just not getting mapped correctly.

EDIT: Looking at the code in the v2 report writer - it's mapping based on the "result" attribute on each test case. I'll have a look and see how it's totting up the values in the console.

**Comment by Almewty** [comment]
I ran into the same problems and decided to look at the code.
I fixed it in my fork [here](https://github.com/Almewty/nunit-v2-framework-driver/commit/e22230f788c83535b8a8ee51be3d35a438ec8fc6) but I am not too sure if it does what it should do because I'm not that familiar with all the States of NUnit.
Feedback would be appreciated!

**Comment by CharliePoole** [comment]
Rather than checking the full `ResultState` you should probably just look at the `TestStatus` component. Its values map directly to the four attributes that need to be set.

**Comment by Almewty** [comment]
Thanks for the advice!
I didn't really find a way to get access to the `TestStatus` component so I at least used the same logic to determine the state [here](https://github.com/Almewty/nunit-v2-framework-driver/commit/be1f561ec7a1c77e23f796d419abd45d4a3d023c). The only way I found was via `TestContext.Result.State`. But I could not find a way to get a `TestContext`.
But now that I know that there is a place to get this information from I would rather find a proper way.
Any advice?

**Comment by CharliePoole** [comment]
AH! Apologies... I forgot you are dealing with a V2 `ResultState`, which is nothing but an enum. In NUnit 3, that becomes a class with Status, Label and Site properties.

I think you should handle them consistent with how we translate to NUnit 3 result states in the `GetTranslatedResultState` method. I think you have done that with the sole exception of `NotRunnable`, which NUnit 3 treats as failed.

It would also be cool if you could avoid recursing through all the children four times and do it just once, but I don't see that as essential to the PR.

**Comment by CharliePoole** [comment]
Fixed by PR #23
